src/internal/ClickListener.js

```diff
@@ -1,14 +1,25 @@
 import PropTypes from 'prop-types';
-/* global document */
-
 import React from 'react';
 
-class ClickListener extends React.Component {
+/**
+ * Generic component used for reacting to a click event happening outside of a
+ * given `children` element.
+ */
+export default class ClickListener extends React.Component {
   static propTypes = {
-    children: PropTypes.node.isRequired,
+    children: PropTypes.element.isRequired,
     onClickOutside: PropTypes.func.isRequired,
   };
 
+  constructor(props) {
+    super(props);
+    // We manually bind handlers in this Component, versus using class
+    // properties, so that we can properly test the `handleRef` handler with
+    // enzyme.
```

Comment:
Curious what the issue was with Enzyme?

Reply:
With class properties, the methods are only available on the instance versus the prototype, so we can't spy on it before creation time 😞

Reply:
Got it @joshblack - Thank you for explaining. Would this work? https://stackoverflow.com/a/39819539/1842550

Reply:
We totally could, would prefer to not extend React classes though.

Reply:
@joshblack You are right that extending React component class is not a good practice. The link suggests only extending the class in tests 😉

Reply:
Haha, totally understand 😄  If the `bind` in constructor is too much, can definitely add that in!

Reply:
Cool @joshblack thank you for understanding - Would be great to avoid having to change our code for the sake of tests, especially if we have a way to have tools/tactics in test solve the issue. (Note: It doesn’t mean we should introduce inherently hard-to-test code like globals, too much React states, etc.) Thanks! 😊

Reply:
The only worry that I have is that the bind pattern is pretty common, whereas the test pattern is obscure. The cost of moving to bind in a constructor seems minor, especially since it's semantic given that we want to inspect a method on a prototype versus on a class instance.

Reply:
@joshblack Fair enough - Though my link was a top hit of https://www.google.co.jp/search?q=class+field+arrow+function+spy+stub+test+mount+react, different people seems to have different opinions: https://github.com/facebook/react/issues/9851#issuecomment-306221157 https://github.com/airbnb/enzyme/issues/365#issuecomment-217368222